packages/core-backend/src/BackendWebSocketService.test.ts

```diff
@@ -0,0 +1,1659 @@
+
+      // Start connection and then advance timers to trigger timeout
+      const connectPromise = service.connect();
+
+      // Handle the promise rejection properly
+      connectPromise.catch(() => {
+        // Expected rejection - do nothing to avoid unhandled promise warning
+      });
+
+      await completeAsyncOperations(TEST_CONSTANTS.TIMEOUT_MS + 50);
+
+      // Now check that the connection failed as expected
+      await expect(connectPromise).rejects.toThrow(
+        `Failed to connect to WebSocket: Connection timeout after ${TEST_CONSTANTS.TIMEOUT_MS}ms`,
+      );
+
+      // Verify we're in error state from the failed connection attempt
+      expect(service.getConnectionInfo().state).toBe(WebSocketState.ERROR);
+
+      const info = service.getConnectionInfo();
+      expect(info.lastError).toContain(
+        `Connection timeout after ${TEST_CONSTANTS.TIMEOUT_MS}ms`,
+      );
+
+      cleanup();
+    });
+  });
+
+  // =====================================================
+  // DISCONNECT TESTS
+  // =====================================================
+  describe('disconnect', () => {
+    it('should disconnect successfully when connected', async () => {
+      const { service, completeAsyncOperations, cleanup } =
+        setupBackendWebSocketService();
+
+      const connectPromise = service.connect();
+      await completeAsyncOperations();
+      await connectPromise;
+
+      await service.disconnect();
+
+      expect(service.getConnectionInfo().state).toBe(
+        WebSocketState.DISCONNECTED,
+      );
+
+      cleanup();
+    }, 10000);
+
+    it('should handle disconnect when already disconnected', async () => {
+      const { service, completeAsyncOperations, cleanup } =
+        setupBackendWebSocketService();
+
+      // Wait for initialization
+      await completeAsyncOperations();
+
+      // Already disconnected - should not throw
+      expect(() => service.disconnect()).not.toThrow();
+
+      expect(service.getConnectionInfo().state).toBe(
+        WebSocketState.DISCONNECTED,
+      );
+
+      cleanup();
+    }, 10000);
+  });
+
+  // =====================================================
+  // SUBSCRIPTION TESTS
+  // =====================================================
+  describe('subscribe', () => {
+    it('should subscribe to channels successfully', async () => {
+      const { service, completeAsyncOperations, getMockWebSocket, cleanup } =
+        setupBackendWebSocketService();
+
+      // Connect first
+      const connectPromise = service.connect();
+      await completeAsyncOperations();
+      await connectPromise;
+
+      const mockCallback = jest.fn();
+      const mockWs = getMockWebSocket();
+
+      // Start subscription
+      const subscriptionPromise = service.subscribe({
+        channels: [TEST_CONSTANTS.TEST_CHANNEL],
+        callback: mockCallback,
+      });
+
+      // Wait for the subscription request to be sent
+      await completeAsyncOperations();
```

Comment:
We seem to be calling this helper a _lot_ in these tests. Is it just that things are continually happening out-of-band? I'm just curious if there is a way we can set up the test so that we can control what happens when more easily. Perhaps there's an event we can wait for and then resolve a promise, rather than wait for the next thing in the promise queue to run. What are your thoughts?